File: src/main/java/org/cerion/stocklist/indicators/Vortex.java
package org.cerion.stocklist.indicators;

import org.cerion.stocklist.PriceList;
import org.cerion.stocklist.arrays.FloatArray;
import org.cerion.stocklist.arrays.PairArray;
import org.cerion.stocklist.arrays.ValueArray;
import org.cerion.stocklist.functions.types.Indicator;

public class Vortex extends IndicatorBase {

    public Vortex() {
        super(Indicator.VORTEX, 14);
    }

    public Vortex(Number ...params) {
        this();
        setParams(params);
    }

    @Override
    public String getName() {
        return "Vortex";
    }

    @Override
    public PairArray eval(PriceList list) {
        return vortex(list, getInt(0) );
    }

    private static PairArray vortex(PriceList list, int period) {
        int size = list.size();
        FloatArray posVI = new FloatArray(size);
        FloatArray negVI = new FloatArray(size);

        float[][] vm = new float[size][2]; // +VM/-VM

        for(int i = 1; i < size; i++) {
            vm[i][0] = Math.abs(list.high(i) - list.low(i-1));
            vm[i][1] = Math.abs(list.low(i) - list.high(i-1));
        }

        // Start at 1 since that is the average value
        posVI.mVal[0] = 1;
        negVI.mVal[0] = 1;

        for(int i = 1; i < size; i++) {
            int count = ValueArray.maxPeriod(i, period);
            float vip = 0;
            float vin = 0;
            float tr = 0;
            for(int j = i - count + 1; j <= i; j++)
            {
                vip += vm[j][0];
                vin += vm[j][1];
                tr += list.tr(j);
            }

            posVI.mVal[i] = vip / tr;
            negVI.mVal[i] = vin / tr;
        }

        return new PairArray(posVI, negVI);
    }

}

File: src/main/java/org/cerion/stocklist/indicators/Vortex.kt
package org.cerion.stocklist.indicators

import org.cerion.stocklist.PriceList
import org.cerion.stocklist.arrays.FloatArray
import org.cerion.stocklist.arrays.PairArray
import org.cerion.stocklist.arrays.ValueArray
import org.cerion.stocklist.functions.types.Indicator

class Vortex() : IndicatorBase(Indicator.VORTEX, 14) {

    constructor(vararg params: Number) : this() {
        setParams(*params)
    }

    override fun getName(): String {
        return "Vortex"
    }

    override fun eval(list: PriceList): PairArray {
        return vortex(list, getInt(0))
    }

    private fun vortex(list: PriceList, period: Int): PairArray {
        val size = list.size
        val posVI = FloatArray(size)
        val negVI = FloatArray(size)

        val vm = Array(size) { kotlin.FloatArray(2) } // +VM/-VM

        for (i in 1 until size) {
            vm[i][0] = Math.abs(list.high(i) - list.low(i - 1))
            vm[i][1] = Math.abs(list.low(i) - list.high(i - 1))
        }

        // Start at 1 since that is the average value
        posVI.mVal[0] = 1f
        negVI.mVal[0] = 1f

        for (i in 1 until size) {
            val count = ValueArray.maxPeriod(i, period)
            var vip = 0f
            var vin = 0f
            var tr = 0f
            for (j in i - count + 1..i) {
                vip += vm[j][0]
                vin += vm[j][1]
                tr += list.tr(j)
            }

            posVI.mVal[i] = vip / tr
            negVI.mVal[i] = vin / tr
        }

        return PairArray(posVI, negVI)
    }

}
